Social login refactor

revant opened the pull request:
Social Login Key converted to `DocType` from `Single DocType`

![slk](https://user-images.githubusercontent.com/1040161/33061389-f831d568-cec1-11e7-8fc4-1e86f3bdb2df.gif)

Screenshot of DocType

![slk-dt](https://user-images.githubusercontent.com/1040161/33061440-24dede08-cec2-11e7-8123-338bc7d114ca.png)

### Discussion

- There were fields for each provider (e.g. `github_username`, `github_userid`) on User. Moved it to custom field. It adds custom field on user for each provider. Should this be done?
- Adding a social login key for a provider adds it's own module with code for parsing authentication_code. i.e adding provider as developer mode it will make changes to files.
- How to move data from single doctype to normal doctype in patch?
    - {provider}.json is migrated into db in model.sync
    - patch.txt runs before sync, need to move client_id and client_secret only after provider.json is migrated. (just `frappe.reload_doc` not enough, need patching or importing `{provider}.json` record)

revant committed:
Added DocType Social Login Key

WIP for https://github.com/frappe/frappe/issues/4496
added basic fields
after_insert add provider_username and provider_userid fields on User dt
on_trash deletes added fields on User dt

revant committed:
Added field to store fontawesome icon for provider

revant committed:
[Patch] Social Login Keys to Social Login Key

revant committed:
[Patch] Social Login Keys to Social Login Key

revant committed:
Merge 'develop' of frappe/frappe into social_login_refactor

revant committed:
Social Login Key generates boilerplate

revant committed:
patch fixed for social_login_refactor

revant committed:
removed patch-not working

revant committed:
use social login keys to initiate flow

revant committed:
Login page shows Social Login Key

revant committed:
Merge 'develop' of frappe/frappe into social_login_refactor

revant committed:
show login via if base_url present

revant committed:
Merge 'develop' of frappe/frappe into social_login_refactor

rmehta commented:
We can have preset configs for standard providers like Google, GitHub, Facebook etc. that the user can select via a drop down. Similar to how it is done for Email Account

Use can also setup their own end points etc as needed.

revant commented:
- [ ] Add presets instead of Standard JSON
- [ ] Use child table on user instead of custom fields for `provider_username` and `provider_userid`.

revant committed:
Merge 'develop' of frappe/frappe into social_login_refactor

revant committed:
removed boilerplate generator

revant committed:
Multiple Changes

fix zxcvbn import in password_strength.py
use of child table instead of additional fields on user dt to store username and userid

revant committed:
Fetched Template on Client JS

revant committed:
Frappe social login template working

revant committed:
Added Social Login Key Templates

revant committed:
Codacy fixes and validate social login key urls

revant committed:
[Patch] Social Login Keys (untested)

revant committed:
Merge 'develop' of frappe/frappe into social_login_refactor

revant committed:
[Fix] Patch refactor social login keys

revant committed:
[Fix] Patch refactor_social_login_keys manually tested

revant committed:
Refactor OAuth 2.0 related changes for Social Login Key

revant committed:
[Fix] Patch refactor social login keys

revant committed:
Merge branch frappe/frappe into social_login_refactor

revant committed:
Test - Adding Frappe Social Login Key

revant committed:
Social Login Key Tests

check added child table entry on user for provider frappe
it also checks if userid is created

revant committed:
[WIP] Office 365 Social Login Key Template

revant committed:
[Fix] Social Login - Redirect URL

revant committed:
Merge 'develop' of frappe/frappe into social_login_refactor

revant committed:
[Test] Single sign-on icons for added provider

revant committed:
Merge 'develop' of frappe/frappe into social_login_refactor

revant committed:
[Fix] Codacy Errors

revant committed:
[Fix] Social Login Key Form JS

revant committed:
Merge 'develop' of frappe/frappe into social_login_refactor

revant committed:
Docs Added for Social Login Key

revant committed:
Merge 'develop' of frappe/frappe into social_login_refactor

revant committed:
[Fix] Patch Refactor Social Login Keys

revant committed:
Merged upstream and resolved conflict

revant committed:
Merge 'develop' of frappe/frappe into social_login_refactor

netchampfaris commented:
- [ ] Handle different icon types (image, icon, emoji) with just `icon` field
- [ ] Move `Enable` field to top
- [ ] Fields which are not editable should be collapsed
https://github.com/frappe/erpnext/wiki/Form-Design-Guidelines
- [ ] Create `.json` files when new Social Key is added, so that they are synced on migrate (?)
- [ ] Move the login methods to a new file (`oauth_logins.py` ?)

revant committed:
Handle different icon types

Handle different icon types (image, icon, emoji) with just icon field

revant committed:
Merge 'develop' of frappe/frappe into social_login_refactor

revant committed:
Move the login methods to a new py file

frappe.integrations.oauth2_logins added
copied whitelisted guest oauth2 redirect endpoints from login.py
removing the functions from login.py will break backward compatibility

revant committed:
Social Login Key Form Changes

Moved Enable field to top
Fields which are not editable are collapsed

revant committed:
Merge 'develop' of frappe/frappe into social_login_refactor

revant commented:
@netchampfaris 
> Create .json files when new Social Key is added, so that they are synced on migrate (?)

Was using that approach initially https://github.com/frappe/frappe/pull/4519/commits/75f7a6a38e3083faa696cea6df903455ea441708

Now using Template approach 
> Similar to how it is done for Email Account

revant committed:
[Fix] Codacy Errors

revant committed:
Corrected Docs, sync.py

revant committed:
Merge 'develop' of frappe/frappe into social_login_refactor

revant committed:
[Docs] Adding a social login provider

revant committed:
[Fix] set frappe userid from User Social Login

revant committed:
[Fix] frappe userid in oauth.py

revant committed:
resolved merge conflict patches.txt

revant committed:
removed icon_type

revant committed:
resolved merge conflict in user.json

Chat fields caused conflict

revant committed:
Merge 'develop' of frappe/frappe into social_login_refactor

netchampfaris committed:
Use frappe.utils.is_image

kickapoo commented:
patch failure

```python
Executing frappe.patches.v10_0.refactor_social_login_keys in monitor.local (11b6bdeee76842cb)
Traceback (most recent call last):
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 174, in _run_module_as_main
    "__main__", fname, loader, pkg_name)
  File "/usr/local/Cellar/python/2.7.13/Frameworks/Python.framework/Versions/2.7/lib/python2.7/runpy.py", line 72, in _run_code
    exec code in run_globals
  File "/Users/kickapoo/frappe-bench/apps/frappe/frappe/utils/bench_helper.py", line 94, in <module>
    main()
  File "/Users/kickapoo/frappe-bench/apps/frappe/frappe/utils/bench_helper.py", line 18, in main
    click.Group(commands=commands)(prog_name='bench')
  File "/Users/kickapoo/frappe-bench/env/lib/python2.7/site-packages/click/core.py", line 722, in __call__
    return self.main(*args, **kwargs)
  File "/Users/kickapoo/frappe-bench/env/lib/python2.7/site-packages/click/core.py", line 697, in main
    rv = self.invoke(ctx)
  File "/Users/kickapoo/frappe-bench/env/lib/python2.7/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/kickapoo/frappe-bench/env/lib/python2.7/site-packages/click/core.py", line 1066, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/Users/kickapoo/frappe-bench/env/lib/python2.7/site-packages/click/core.py", line 895, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/Users/kickapoo/frappe-bench/env/lib/python2.7/site-packages/click/core.py", line 535, in invoke
    return callback(*args, **kwargs)
  File "/Users/kickapoo/frappe-bench/env/lib/python2.7/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/Users/kickapoo/frappe-bench/apps/frappe/frappe/commands/__init__.py", line 24, in _func
    ret = f(frappe._dict(ctx.obj), *args, **kwargs)
  File "/Users/kickapoo/frappe-bench/apps/frappe/frappe/commands/site.py", line 222, in migrate
    migrate(context.verbose, rebuild_website=rebuild_website)
  File "/Users/kickapoo/frappe-bench/apps/frappe/frappe/migrate.py", line 31, in migrate
    frappe.modules.patch_handler.run_all()
  File "/Users/kickapoo/frappe-bench/apps/frappe/frappe/modules/patch_handler.py", line 29, in run_all
    if not run_single(patchmodule = patch):
  File "/Users/kickapoo/frappe-bench/apps/frappe/frappe/modules/patch_handler.py", line 63, in run_single
    return execute_patch(patchmodule, method, methodargs)
  File "/Users/kickapoo/frappe-bench/apps/frappe/frappe/modules/patch_handler.py", line 83, in execute_patch
    frappe.get_attr(patchmodule.split()[0] + ".execute")()
  File "/Users/kickapoo/frappe-bench/apps/frappe/frappe/patches/v10_0/refactor_social_login_keys.py", line 49, in execute
    social_login_keys = frappe.get_doc("Social Login Keys", "Social Login Keys")
  File "/Users/kickapoo/frappe-bench/apps/frappe/frappe/__init__.py", line 638, in get_doc
    return frappe.model.document.get_doc(*args, **kwargs)
  File "/Users/kickapoo/frappe-bench/apps/frappe/frappe/model/document.py", line 66, in get_doc
    controller = get_controller(doctype)
  File "/Users/kickapoo/frappe-bench/apps/frappe/frappe/model/base_document.py", line 36, in get_controller
    module = load_doctype_module(doctype, module_name)
  File "/Users/kickapoo/frappe-bench/apps/frappe/frappe/modules/utils.py", line 187, in load_doctype_module
    raise ImportError('Module import failed for {0} ({1})'.format(doctype, module_name + ' Error: ' + str(e)))
ImportError: Module import failed for Social Login Keys (frappe.integrations.doctype.social_login_keys.social_login_keys Error: No module named social_login_keys.social_login_keys)
```

revant commented:
https://github.com/frappe/frappe/pull/4749
